kolibri/plugins/user/assets/src/views/SignInPage/index.vue

```diff
@@ -286,7 +286,9 @@
       },
     },
     watch: {
-      username: 'setSuggestionTerm',
+      username() {
```

Comment:
Hrm, rather than forcing this syntax, oughtn't we just add an additional check as to whether the method is used as a watch callback by string reference?

Reply:
I don't particularly mind forcing the more verbose function syntax. Referencing functions by string value seems unnecessarily magical and [isn't covered in the Vue intro guide](https://vuejs.org/v2/guide/computed.html).

Opaque without knowledge of Vue:

```javascript
    watch: {
      username: 'setSuggestionTerm',
    }
```

Very explicit and readable:

```javascript
    watch: {
      username() {
        this.setSuggestionTerm();
      },
    }
```

Concise and explicit, but maybe less readable:

```javascript
    watch: {
      username: () => this.setSuggestionTerm(),
    }
```

Reply:
Yep, as described in the commit message, I didn't find any reference (not even here https://vuejs.org/v2/api/#vm-watch, have I missed something?) and that's why I decided to update the code instead of adding support to the rule.

Reply:
It's under the component Options section: https://vuejs.org/v2/api/#watch (not the watch method itself), second item in the example.

Reply:
Ah, sorry, when I posted the link I noticed I'm in the wrong place https://vuejs.org/v2/api/#watch.

Reply:
And I obviously checked the wrong one before. No problem to add the support for this then.

Reply:
And from that same reference @indirectlylit:
```
Note that you should not use an arrow function to define a watcher (e.g. searchQuery: newValue => this.updateAutocomplete(newValue)). The reason is arrow functions bind the parent context, so this will not be the Vue instance as you expect and this.updateAutocomplete will be undefined.
```

Reply:
Also, if the method is reading the newValue (and any of the other values passed into the watcher function), then the object method form becomes even more verbose:
```
    watch: {
      username(newVal, oldVal) {
        this.setSuggestionTerm(newVal, oldVal);
      },
    }
```

Reply:
Yeah it is indeed and I like it because you immediately know what's happening there, but I am quite into functions so it's just a personal preference. No problem to support more syntaxes.

Reply:
A few other considerations:

* `watch` is a code smell: new Vue devs will often use it when `computed` would have been more appropriate
* In Kolibri we use `this.$watch(...)` 5x and `watch: { ... }` 32x.
* We use the string syntactic sugar in just 4 components, and one of them actually [looks like a bug](https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/learn/assets/src/views/TotalPoints.vue#L40) that would have been caught if it were written using the full function syntax.

I vote to ban :)

Reply:
> We use the string syntactic sugar in just 4 components, and one of them actually looks like a bug that would have been caught if it were written using the full function syntax.

After checking that reference in our codebase, this actually sounds like a pretty good reason to me.

Reply:
If we ban it, the question is - should this particular rule be responsible for this? It might be confusing for folks who know how to use the string syntax but don't know about this decision to not support it in our codebase. Because in this situation, they would be basically using valid Vue syntax but the report would say "Unused method found...". I was thinking about adding more help text but it's long enough and anyway, this sounds like something to be forced with another rule?

Reply:
I think banning should be another rule, yes, so making this rule work for the banned case seems reasonable, then adding a very lightweight 'vue-watch-no-string' rule seems sensible.

Reply:
I am not certain that the bug would have been caught without this as the bug arise because of the mapped action being removed from the component definition and that not being noticed in review. 

I think the main reason it is extant is that it will only ever get triggered right as someone is logged out, which will almost always trigger a redirect.

That said, banning is fine.

Reply:
Could you file a follow up issue for this banning?

Reply:
#5380

Reply:
thanks guys.

The reason I assumed it would have been caught was that I think this would have resulted in an 'undefined function `fetchPoints`' error?

```javascript
watch: {
  currentUserId() { this.fetchPoints(); },
},
```

Reply:
Yes - but Vue would have thrown a similar error when the named watcher function didn't exist - so the reason the bug didn't get caught was because the watcher was only being invoked during a page redirect.